File: 72_llm_market_simulation/python/examples/bubble_detection.py
#!/usr/bin/env python3
"""
Bubble Detection Simulation Example

Demonstrates how momentum-heavy markets can form bubbles
and how the simulation can detect bubble formation.
"""

import sys
import os

# Add parent directory to path for imports
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np
from market import OrderBook, MarketEnvironment
from agents import ValueInvestorAgent, MomentumTraderAgent, MarketMakerAgent
from simulation import SimulationEngine
from simulation.metrics import calculate_performance_metrics, detect_bubble


def main():
    """Run bubble-prone simulation"""
    print("=" * 60)
    print("LLM Market Simulation - Bubble Detection Example")
    print("=" * 60)

    np.random.seed(123)  # For reproducibility

    # Parameters that encourage bubble formation
    initial_price = 100.0
    fundamental_value = 100.0
    num_steps = 300

    print("\nScenario: Momentum-Heavy Market")
    print("Creating market with more momentum traders than value investors...")
    print("This configuration is prone to bubble formation.\n")

    # Create simulation engine with higher volatility
    engine = SimulationEngine(
        initial_price=initial_price,
        fundamental_value=fundamental_value,
        volatility=0.025  # Slightly higher volatility
    )

    initial_cash = 100000.0
    initial_shares = 100

    # Only 1 value investor (weak anchor to fundamentals)
    agent = ValueInvestorAgent(
        agent_id="value_1",
        initial_cash=initial_cash,
        initial_shares=initial_shares,
        fundamental_value=fundamental_value,
        discount_threshold=0.15,  # Very patient, only buys at big discounts
        premium_threshold=0.15
    )
    engine.add_agent(agent)
    print(f"Added 1 Value Investor (threshold: 15%)")

    # Many momentum traders (trend followers create positive feedback)
    for i in range(5):
        agent = MomentumTraderAgent(
            agent_id=f"momentum_{i+1}",
            initial_cash=initial_cash,
            initial_shares=initial_shares,
            short_window=3,  # Fast signals
            long_window=8,
            entry_threshold=0.01  # Low threshold, trades often
        )
        engine.add_agent(agent)
    print(f"Added 5 Momentum Traders (aggressive settings)")

    # Market makers
    for i in range(2):
        agent = MarketMakerAgent(
            agent_id=f"mm_{i+1}",
            initial_cash=initial_cash * 2,
            initial_shares=initial_shares * 2,
            base_spread=0.002
        )
        engine.add_agent(agent)
    print(f"Added 2 Market Makers")

    print(f"\nTotal: 1 Value + 5 Momentum + 2 MM = 8 agents")
    print("Momentum traders outnumber value investors 5:1")

    # Run simulation
    print(f"\nRunning simulation for {num_steps} steps...")
    result = engine.run(num_steps=num_steps, verbose=False)

    # Analyze for bubbles
    print("\n" + "=" * 60)
    print("BUBBLE ANALYSIS")
    print("=" * 60)

    bubble_info = detect_bubble(
        result.price_history,
        result.fundamental_history,
        bubble_threshold=0.30  # 30% above fundamental = bubble
    )

    print(f"\nBubble Threshold: 30% above fundamental value")
    print(f"Bubble Detected: {bubble_info.get('bubble_detected', False)}")

    if bubble_info.get("bubble_detected"):
        print(f"Number of Bubble Periods: {bubble_info.get('num_bubbles', 0)}")
        print(f"Maximum Price Deviation: {bubble_info.get('max_deviation', 0)*100:.1f}%")
        print(f"Time in Bubble: {bubble_info.get('time_in_bubble_pct', 0):.1f}% of simulation")

        print("\nBubble Periods:")
        for i, period in enumerate(bubble_info.get("bubble_periods", [])):
            print(f"  Period {i+1}:")
            print(f"    Start Step: {period['start']}")
            print(f"    Peak Step: {period['peak']}")
            print(f"    End Step: {period['end']}")
            print(f"    Duration: {period['duration']} steps")
            print(f"    Peak Deviation: {period['peak_deviation']*100:.1f}%")
            peak_price = result.price_history[period['peak']]
            peak_fund = result.fundamental_history[period['peak']]
            print(f"    Peak Price: ${peak_price:.2f} vs Fundamental: ${peak_fund:.2f}")
    else:
        print("No significant bubble formed in this simulation run.")
        print("(Try running multiple times - bubbles are probabilistic)")

    # Price statistics
    print("\n" + "=" * 60)
    print("PRICE STATISTICS")
    print("=" * 60)

    prices = np.array(result.price_history)
    fundamentals = np.array(result.fundamental_history)
    deviations = (prices - fundamentals) / fundamentals * 100

    print(f"\nPrice Range:")
    print(f"  Minimum: ${prices.min():.2f} ({deviations.min():.1f}% from fundamental)")
    print(f"  Maximum: ${prices.max():.2f} ({deviations.max():.1f}% from fundamental)")
    print(f"  Final: ${prices[-1]:.2f} ({deviations[-1]:.1f}% from fundamental)")

    print(f"\nDeviation Statistics:")
    print(f"  Mean Deviation: {deviations.mean():.1f}%")
    print(f"  Std Deviation: {deviations.std():.1f}%")
    print(f"  Time Above +20%: {(deviations > 20).mean()*100:.1f}%")
    print(f"  Time Below -20%: {(deviations < -20).mean()*100:.1f}%")

    # Performance metrics
    metrics = calculate_performance_metrics(result.price_history, result.fundamental_history)

    print("\n" + "=" * 60)
    print("MARKET PERFORMANCE")
    print("=" * 60)
    print(f"  Total Return: {metrics.get('total_return_pct', 0):.2f}%")
    print(f"  Volatility: {metrics.get('volatility_pct', 0):.2f}%")
    print(f"  Sharpe Ratio: {metrics.get('sharpe_ratio', 0):.3f}")
    print(f"  Max Drawdown: {metrics.get('max_drawdown_pct', 0):.2f}%")
    print(f"  Fundamental Correlation: {metrics.get('fundamental_correlation', 0):.3f}")

    # Simple ASCII visualization
    print("\n" + "=" * 60)
    print("PRICE CHART (ASCII)")
    print("=" * 60)

    # Sample every nth point for display
    sample_rate = max(1, len(prices) // 50)
    sampled_prices = prices[::sample_rate]
    sampled_funds = fundamentals[::sample_rate]

    min_p = min(min(sampled_prices), min(sampled_funds))
    max_p = max(max(sampled_prices), max(sampled_funds))
    height = 15
    width = len(sampled_prices)

    def scale(val):
        if max_p == min_p:
            return height // 2
        return int((val - min_p) / (max_p - min_p) * (height - 1))

    chart = [[' ' for _ in range(width)] for _ in range(height)]

    # Plot prices
    for x, p in enumerate(sampled_prices):
        y = height - 1 - scale(p)
        chart[y][x] = '*'

    # Plot fundamentals
    for x, f in enumerate(sampled_funds):
        y = height - 1 - scale(f)
        if chart[y][x] == ' ':
            chart[y][x] = '-'

    print(f"${max_p:.0f} |", end="")
    for row in chart:
        print("".join(row))
        print("      |", end="")
    print(f"\n${min_p:.0f} |" + "_" * width)
    print("       " + "Step 0" + " " * (width - 12) + f"Step {num_steps}")
    print("\nLegend: * = Price, - = Fundamental Value")

    print("\nDone!")


if __name__ == "__main__":
    main()

File: 72_llm_market_simulation/rust_market_sim/examples/bubble_detection.rs
//! Bubble Detection Example
//!
//! Demonstrates how momentum-heavy markets can form bubbles
//! and how to detect them using simulation metrics.

use llm_market_sim::{
    agents::{ValueInvestor, MomentumTrader, MarketMaker},
    simulation::{SimulationEngine, calculate_performance_metrics, detect_bubble},
};

fn main() {
    println!("Bubble Detection Simulation Example");
    println!("====================================");
    println!();

    println!("Scenario: Momentum-Heavy Market");
    println!("This configuration is prone to bubble formation due to");
    println!("the dominance of trend-following traders over value investors.");
    println!();

    // Create simulation with settings that encourage bubbles
    let mut engine = SimulationEngine::new(100.0, 100.0, 0.025);

    // Only 1 value investor (weak anchor to fundamentals)
    engine.add_agent(Box::new(
        ValueInvestor::new("value_1".to_string(), 100000.0, 100, 100.0)
            .with_thresholds(0.15, 0.15) // Very patient
    ));
    println!("Added: 1 Value Investor (threshold: 15%)");

    // Many aggressive momentum traders
    for i in 0..5 {
        engine.add_agent(Box::new(
            MomentumTrader::new(
                format!("momentum_{}", i + 1),
                100000.0,
                100,
            )
            .with_windows(3, 8) // Fast signals
            .with_threshold(0.01) // Low threshold
        ));
    }
    println!("Added: 5 Momentum Traders (aggressive settings)");

    // Market makers
    for i in 0..2 {
        engine.add_agent(Box::new(
            MarketMaker::new(format!("mm_{}", i + 1), 200000.0, 200)
        ));
    }
    println!("Added: 2 Market Makers");
    println!();

    println!("Running simulation for 300 steps...");
    let result = engine.run(300);
    println!("Done!");
    println!();

    // Analyze for bubbles
    println!("═══════════════════════════════════════");
    println!("              BUBBLE ANALYSIS          ");
    println!("═══════════════════════════════════════");
    println!();

    let bubble_info = detect_bubble(&result.price_history, &result.fundamental_history, 0.30);

    println!("Bubble Threshold: 30% above fundamental");
    println!("Bubble Detected: {}", bubble_info.bubble_detected);
    println!("Number of Bubbles: {}", bubble_info.num_bubbles);
    println!("Max Deviation: {:.1}%", bubble_info.max_deviation * 100.0);
    println!("Time in Bubble: {:.1}%", bubble_info.time_in_bubble_pct);

    if !bubble_info.bubble_periods.is_empty() {
        println!();
        println!("Bubble Periods:");
        for (i, period) in bubble_info.bubble_periods.iter().enumerate() {
            println!("  Period {}:", i + 1);
            println!("    Start: step {}", period.start);
            println!("    Peak: step {} ({:.1}% deviation)", period.peak, period.peak_deviation * 100.0);
            println!("    End: step {}", period.end);
            println!("    Duration: {} steps", period.duration);
            if period.peak < result.price_history.len() {
                println!(
                    "    Peak Price: ${:.2} vs Fundamental: ${:.2}",
                    result.price_history[period.peak],
                    result.fundamental_history[period.peak]
                );
            }
        }
    }

    println!();
    println!("═══════════════════════════════════════");
    println!("            PRICE STATISTICS           ");
    println!("═══════════════════════════════════════");
    println!();

    let prices = &result.price_history;
    let fundamentals = &result.fundamental_history;

    let min_price = prices.iter().cloned().fold(f64::INFINITY, f64::min);
    let max_price = prices.iter().cloned().fold(f64::NEG_INFINITY, f64::max);

    let deviations: Vec<f64> = prices.iter()
        .zip(fundamentals.iter())
        .map(|(p, f)| (p - f) / f * 100.0)
        .collect();

    let min_dev = deviations.iter().cloned().fold(f64::INFINITY, f64::min);
    let max_dev = deviations.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    let mean_dev: f64 = deviations.iter().sum::<f64>() / deviations.len() as f64;

    println!("Price Range:");
    println!("  Minimum: ${:.2} ({:.1}% from fundamental)", min_price, min_dev);
    println!("  Maximum: ${:.2} ({:.1}% from fundamental)", max_price, max_dev);
    println!("  Final: ${:.2}", result.final_price);
    println!();
    println!("Deviation Statistics:");
    println!("  Mean: {:.1}%", mean_dev);
    let above_20 = deviations.iter().filter(|&&d| d > 20.0).count() as f64 / deviations.len() as f64 * 100.0;
    let below_20 = deviations.iter().filter(|&&d| d < -20.0).count() as f64 / deviations.len() as f64 * 100.0;
    println!("  Time Above +20%: {:.1}%", above_20);
    println!("  Time Below -20%: {:.1}%", below_20);

    println!();
    println!("═══════════════════════════════════════");
    println!("          PERFORMANCE METRICS          ");
    println!("═══════════════════════════════════════");
    println!();

    let metrics = calculate_performance_metrics(
        prices,
        Some(fundamentals),
        0.02,
        252.0,
    );

    println!("Total Return: {:.2}%", metrics.total_return_pct);
    println!("Volatility: {:.2}%", metrics.volatility * 100.0);
    if let Some(sharpe) = metrics.sharpe_ratio {
        println!("Sharpe Ratio: {:.3}", sharpe);
    }
    println!("Max Drawdown: {:.2}%", metrics.max_drawdown * 100.0);
    if let Some(corr) = metrics.fundamental_correlation {
        println!("Fundamental Correlation: {:.3}", corr);
    }

    println!();
    println!("Done!");
}
